src/audio/smart_amp/smart_amp.c

```diff
@@ -533,45 +484,72 @@ static int smart_amp_trigger(struct comp_dev *dev, int cmd)
 	return ret;
 }
 
-static int smart_amp_process(struct comp_dev *dev,
-			     const struct audio_stream __sparse_cache *source,
-			     const struct audio_stream __sparse_cache *sink,
-			     uint32_t frames, int8_t *chan_map,
-			     bool is_feedback)
+static int smart_amp_ff_process(struct comp_dev *dev,
+				const struct audio_stream __sparse_cache *source,
+				const struct audio_stream __sparse_cache *sink,
+				uint32_t frames, const int8_t *chan_map)
 {
 	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct smart_amp_mod_data_base *mod = sad->mod_data;
 	int ret;
 
-	if (!is_feedback)
-		ret = smart_amp_ff_copy(dev, frames,
-					source, sink,
-					chan_map, sad->mod_handle,
-					sad->in_channels, sad->out_channels);
-	else
-		ret = smart_amp_fb_copy(dev, frames,
-					source, sink,
-					chan_map, sad->mod_handle,
-					audio_stream_get_channels(source));
-	return ret;
+	sad->ff_mod.consumed = 0;
+	sad->out_mod.produced = 0;
+
+	if (frames == 0) {
+		comp_warn(dev, "smart_amp_copy(): feed forward frame size zero warning.");
+		return 0;
+	}
+
+	if (frames > SMART_AMP_FF_BUF_DB_SZ) {
+		comp_err(dev, "smart_amp_copy(): feed forward frame size overflow: %u", frames);
+		sad->ff_mod.consumed = frames;
+		return -EINVAL;
+	}
+
+	sad->ff_get_frame(&sad->ff_mod, frames, source, chan_map);
+
+	ret = mod->mod_ops->ff_proc(mod, frames, &sad->ff_mod, &sad->out_mod);
+	if (ret) {
+		comp_err(dev, "smart_amp_copy(): feed forward inner model process error");
+		return ret;
+	}
+
+	sad->ff_set_frame(&sad->out_mod, sad->out_mod.produced, sink);
+
+	return 0;
 }
 
-static smart_amp_proc get_smart_amp_process(struct comp_dev *dev)
+static int smart_amp_fb_process(struct comp_dev *dev,
+				const struct audio_stream __sparse_cache *source,
+				uint32_t frames, const int8_t *chan_map)
 {
 	struct smart_amp_data *sad = comp_get_drvdata(dev);
-	struct comp_buffer __sparse_cache *source_buf = buffer_acquire(sad->source_buf);
-	enum sof_ipc_frame fmt = audio_stream_get_frm_fmt(&source_buf->stream);
+	struct smart_amp_mod_data_base *mod = sad->mod_data;
+	int ret;
 
-	buffer_release(source_buf);
+	sad->fb_mod.consumed = 0;
 
-	switch (fmt) {
-	case SOF_IPC_FRAME_S16_LE:
-	case SOF_IPC_FRAME_S24_4LE:
-	case SOF_IPC_FRAME_S32_LE:
-		return smart_amp_process;
-	default:
-		comp_err(dev, "smart_amp_process() error: not supported frame format");
-		return NULL;
+	if (frames == 0) {
+		comp_warn(dev, "smart_amp_copy(): feedback frame size zero warning.");
+		return 0;
```

Comment:
this warning message is as it was actually (just moved here due to the revamp.

Based on my previous debugging experience, this can be reproduced every now and then with no harm (well... at least as far as I've met), but is not an expected behavior.